krun/krun.c

```diff
@@ -0,0 +1,227 @@
+#define EN1_OS      1 << 4
+// Enable counting in higer rings
+#define EN1_USR     1 << 5
+// Enable counting for all core threads (if any)
+#define EN1_ANYTHR  1 << 6
+
+/* For passing args to krun_read_msrs() */
+struct krun_msr_args {
+	u64 aperf;
+	u64 mperf;
+	u64 ctr1;
+};
+
+/*
+ * Take low-latency measurements -- in start order -- for the current CPU core.
+ *
+ * For speed: no branching, no calls.
+ */
+void krun_read_msrs_start(void *data)
+{
+	u32 hi, lo;
+	struct krun_msr_args *args = (struct krun_msr_args *) data;
+
+	asm volatile(
+	    "rdmsr\n\t"
+	    : "=d" (hi), "=a" (lo)      // out
+	    : "c"(IA32_APERF)           // in
+	    :);                         // clobber
+	args->aperf = ((u64) hi << 32) | lo;
+
+	asm volatile(
+	    "rdmsr\n\t"
+	    : "=d" (hi), "=a" (lo)      // out
+	    : "c"(IA32_MPERF)           // in
+	    : );                        // clobber
+	args->mperf = ((u64) hi << 32) | lo;
+
+	asm volatile(
+	    "rdmsr\n\t"
+	    : "=d" (hi), "=a" (lo)      // out
+	    : "c"(IA32_PERF_FIXED_CTR1) // in
+	    :);                         // clobber
+	args->ctr1 = ((u64) hi << 32) | lo;
+}
+
+/*
+ * Same as krun_read_msrs_start(), but measurements taken in stop order.
+ */
+void krun_read_msrs_stop(void *data)
+{
+	u32 hi, lo;
+	struct krun_msr_args *args = (struct krun_msr_args *) data;
+
+	asm volatile(
+	    "rdmsr\n\t"
+	    : "=d" (hi), "=a" (lo)      // out
+	    : "c"(IA32_PERF_FIXED_CTR1) // in
+	    :);                         // clobber
+	args->ctr1 = ((u64) hi << 32) | lo;
+
+	asm volatile(
+	    "rdmsr\n\t"
+	    : "=d" (hi), "=a" (lo)      // out
+	    : "c"(IA32_APERF)           // in
+	    :);                         // clobber
+	args->aperf = ((u64) hi << 32) | lo;
+
+	asm volatile(
+	    "rdmsr\n\t"
+	    : "=d" (hi), "=a" (lo)      // out
+	    : "c"(IA32_MPERF)           // in
+	    : );                        // clobber
+	args->mperf = ((u64) hi << 32) | lo;
+}
+
+/* Reset the MSRs of interest on the current CPU core */
+void krun_reset_msrs(void *ununsed)
+{
+	asm volatile(
+	    "xor %%eax, %%eax\n\t"
+	    "xor %%edx, %%edx\n\t"
+	    "mov %0, %%ecx\n\t"
+	    "wrmsr\n\t"
+	    "mov %1, %%ecx\n\t"
+	    "wrmsr\n\t"
+	    "mov %2, %%ecx\n\t"
+	    "wrmsr\n\t"
+	    :				// out
+	    : "i"(IA32_APERF), "i"(IA32_MPERF), "i"(IA32_PERF_FIXED_CTR1) // in
+	    :"eax", "ecx", "edx");      // clobber
+}
+
+/* ------------------------
+ * System call entry points
+ * ------------------------
+ */
+
+/*
+ * For the specified `core', read the MSRs and return them via `aperf', `mperf'
+ * and `ctr1'. If `index' is 0, then read the MSRs in start order, otherwise
+ * read them in stop order.
+ *
+ * Returns non-zero on error.
+ */
+asmlinkage int sys_krun_read_msrs(int core, int index, u64 *aperf, u64 *mperf, u64 *ctr1)
+{
+	struct krun_msr_args args;
+	unsigned long err;
+
+	/*
+	 * Force call onto the desired core and then read the MSRs in the
+	 * desired order.
+	 */
+	if (index == 0) {
```

Comment:
Shall we have two system calls, one for start, one for stop, to avoid this branch?